se_apps/include/interface.h

```diff
@@ -22,84 +22,163 @@
-		double current_frame = clockData.tv_sec
-				+ clockData.tv_nsec / 1000000000.0;
-		static double first_frame = current_frame;
-
-		int frame = std::ceil((current_frame - first_frame) * (double) _fps);
-		_frame = frame;
-
-		double tpf = 1.0 / _fps;
-		double ttw = ((double) _frame * tpf - current_frame + first_frame);
-		if (_blocking_read) {
-			if (ttw > 0)
-				usleep(1000000.0 * ttw);
-		}
-
-	}
-
-	bool cameraActive;
-	bool cameraOpen;
-protected:
-	int _frame;
-	int _fps;bool _blocking_read;
+      double current_frame = clockData.tv_sec
+        + clockData.tv_nsec / 1000000000.0;
+      static double first_frame = current_frame;
+
+      int frame = std::ceil((current_frame - first_frame) * (double) _fps);
+      _frame = frame;
+
+      double tpf = 1.0 / _fps;
+      double ttw = ((double) _frame * tpf - current_frame + first_frame);
+      if (_blocking_read) {
+        if (ttw > 0)
+          usleep(1000000.0 * ttw);
+      }
+
+    }
+
+    inline bool readNextPose(Eigen::Matrix4f& pose) {
+      std::string line;
+      while (true) {
+        std::getline(_gt_file,line);
+        // EOF reached
+        if (!_gt_file.good()) {
+          //std::cout << "EOF reached" << std::endl;
+          return false;
+        }
+        // Ignore comment lines
+        if (line[0] == '#') {
+          //std::cout << "Skipping comment" << std::endl;
+          continue;
+        }
+        // Data line read, split on spaces
+        std::vector<std::string> data;
+        splitString(line, ' ', data);
+        size_t N = data.size();
+        if (N < 7) {
+          std::cout << "Invalid ground truth file format."
+            << "Expected line format: ... tx ty tz qx qy qz qw" << std::endl;
+          return false;
+        }
+        // Read the last 7 columns
+        Eigen::Vector3f tran (std::stof(data[N-7]), std::stof(data[N-6]),
+          std::stof(data[N-5]));
+        Eigen::Quaternionf quat (std::stof(data[N-1]), std::stof(data[N-4]),
+          std::stof(data[N-3]), std::stof(data[N-2]));
+        pose = Eigen::Matrix4f::Identity();
+        pose.block<3,3>(0,0) = quat.toRotationMatrix();
+        pose.block<3,1>(0,3) = tran;
+        _pose_num++;
+		// TODO
+		// [0,0,0] is currently a corner of the volume. It is not possible to
+		// set the initial position to the volume center because its dimensions
+		// are not known to the reader.
+        // Subtract the first position from the current position
+        //if (_pose_num == 1) {
+        //  _initialPose = pose;
+        //}
+        //pose.block<3,1>(0,3) -= _initialPose.block<3,1>(0,3);
+
+        // Apply the transform to the pose
+        pose = _transform * pose;
+        return true;
+      }
+    }
+
+    bool cameraActive;
+    bool cameraOpen;
+  protected:
+    int _frame;
+    size_t _pose_num;
+    int _fps;
+    bool _blocking_read;
+    std::string _data_path;
+    std::string _groundtruth_path;
+    std::ifstream _gt_file;
+    Eigen::Matrix4f _initialPose;
```

Comment:
I am not sure, what is the difference between _initialPose and _transform?

Reply:
`_initialPose` was a leftover variable from some commented out code. Removed in d0e0cf4dcde9e2be2904002f599c9a285638de8c. `_transform` is the transform matrix optionally applied to all ground truth poses. It is provided using the `-G` option.

Reply:
Perfect! Thanks!